### DISCUSSION TITLE
Great Changes in version 5! Only having an issue with tokens if someone could help.

### DISCUSSION THREAD

**Discussion by cryoaura** [apioo/fusio#560] [Q&A] [answered]
Just to start, awesome changes on the worker setup. Having the workers in connections is a much better feel and the changes made make everything feel more cohesive.

Everything appears to be working great after upgrading - except when I check out the tokens areas. I get a message that says:
"Access to this operation is not in the scope of the provided token in [removed pathing]vendor/fusio/impl/src/Service/Security/TokenValidator.php on line 160"

This happens if i go to Analytics->Token in the Fusio app or in the new Account app if i click the "Token" tab.

Any idea on how to enable tokens on accounts from an upgrade?

Appreciate everything you all have done, I think the worker re-work was great and really dig being able to make user access tokens rather than just app-based.

**Comment by chriskapp** [comment]
Hi @cryoaura thanks for the great feedback. Regarding your problem, you probably only need to add the "token" scope to your account, which we have added with the new version. So go to the user panel and grant your user account the token scope s.
![image](https://github.com/apioo/fusio/assets/2505846/0227dfa6-db66-426c-8290-1d305d59b148)

**Reply by cryoaura** [reply]
Looks like that is checked as that was one of the first troubleshooting steps. Any ideas of what to check out next?

**Reply by cryoaura** [reply]
Another peculiar thing, it seems like when I try and remove backend.token and consumer.token from a user, when I go back and check the user again, the change never took place.

**Reply by cryoaura** [reply]
In TokenValidator.php if I dump out $accessToken['scope'] for my account in the error, I do see backend.token in the list.

If I do the same with $availableScopes, it says it's an empty array. It seems this gets populated via findByOperationId, and I see a comment where it says there must be at least one scope which explicitly allows the request.

Am I missing a scope that authorizes the route for tokens or was I supposed to make one?

I'm getting the same error on webhooks.

**Reply by cryoaura** [reply]
and where i'm out of ideas, I saw that the token.getAll is a hidden operation, and when I faked the ui to edit that operation, there are no scopes attached to it, possibly by design?

![image](https://github.com/apioo/fusio/assets/45918769/bea45b27-2a84-446d-a1ad-79f28d4291f1)

**Reply by chriskapp** [reply]
ok, to further debug this, could you check the following query

```sql
SELECT * FROM fusio_operation WHERE http_path LIKE '/backend/token';
```

and then check which scopes are assigned to this operation, by using the id from the result s.

```sql
SELECT * FROM fusio_scope_operation WHERE operation_id = [operation_id];
```

This should return one scope "backend.token" which you can then check with:

```sql
SELECT * FROM fusio_scope WHERE id = [scope_id];
```

Through this we could check whether the fitting scopes are assigned to the operation. The webhook is also a new scope which we have added with the last release.

**Reply by cryoaura** [reply]
Ah, I get zero results when looking in "fusio_scope_operation" for those operation IDs. Is it possible the migrate script just simply didn't add these pieces?

**Reply by chriskapp** [reply]
yes indeed the migrate script should have added those scopes. Since the `/backend/token` operation has only one scope you could also get the scope id for the `backend.token` scope and add it to the `fusio_scope_operation` table, this should then fix the problem. Do you have the output of the upgrade command, this would be interesting to better understand why those scopes were not added.

**Reply by cryoaura** [reply]
Alrighty, do you have the default records that should have went in or is it pretty straightforward? I found two operations that started with /backend/token, do I need that on each? And for the webhooks, which is that?

If the migration dumped a log somewhere, I haven't deleted anything - did it have one?

**Reply by chriskapp** [reply, marked as answer]
sure, you could try to use this, but you should check whether the scope and operation ids point to the correct entries, you may have different ids, this is from a local Fusio installation s.

## backend token
```sql
INSERT INTO fusio_scope_operation SET scope_id = 31, operation_id = 135, allow = 1;
INSERT INTO fusio_scope_operation SET scope_id = 31, operation_id = 136, allow = 1;
```

## consumer token
```sql
INSERT INTO fusio_scope_operation SET scope_id = 44, operation_id = 169, allow = 1;
INSERT INTO fusio_scope_operation SET scope_id = 44, operation_id = 170, allow = 1;
INSERT INTO fusio_scope_operation SET scope_id = 44, operation_id = 171, allow = 1;
INSERT INTO fusio_scope_operation SET scope_id = 44, operation_id = 172, allow = 1;
INSERT INTO fusio_scope_operation SET scope_id = 44, operation_id = 173, allow = 1;
```

## backend webhook
```sql
INSERT INTO fusio_scope_operation SET scope_id = 35, operation_id = 147, allow = 1;
INSERT INTO fusio_scope_operation SET scope_id = 35, operation_id = 148, allow = 1;
INSERT INTO fusio_scope_operation SET scope_id = 35, operation_id = 149, allow = 1;
INSERT INTO fusio_scope_operation SET scope_id = 35, operation_id = 150, allow = 1;
INSERT INTO fusio_scope_operation SET scope_id = 35, operation_id = 151, allow = 1;
```

## consumer webhook
```sql
INSERT INTO fusio_scope_operation SET scope_id = 45, operation_id = 174, allow = 1;
INSERT INTO fusio_scope_operation SET scope_id = 45, operation_id = 175, allow = 1;
INSERT INTO fusio_scope_operation SET scope_id = 45, operation_id = 176, allow = 1;
INSERT INTO fusio_scope_operation SET scope_id = 45, operation_id = 177, allow = 1;
INSERT INTO fusio_scope_operation SET scope_id = 45, operation_id = 178, allow = 1;
```

**Reply by cryoaura** [reply]
Awesome! That did the trick. One thing I noticed in the account->token area when generating is it's hard-capped at 365 in the token.php. Is there a way we could have the Max Token Length be a setting in fusio so it's more user-controllable? We tend to set some tokens for a longer duration and if there is a breach or suspected abuse, we simply remove the tokens.

I was thinking an option in Fusio->System->Settings, defaults at 365 but can be changed to whatever is desired. Just curious. Thanks for the help with everything!
